File: src/test/java/io/github/intellij/dlanguage/features/DDocumentationProviderTest.java
package io.github.intellij.dlanguage.features;

import com.intellij.codeInsight.documentation.DocumentationManager;
import com.intellij.psi.PsiElement;
import com.intellij.testFramework.fixtures.BasePlatformTestCase;

public class DDocumentationProviderTest extends BasePlatformTestCase {

    private DDocumentationProvider provider;

    public void setUp() throws Exception {
        super.setUp();

        provider = new DDocumentationProvider();
    }

    @Override
    protected String getTestDataPath() {
        return this.getClass().getClassLoader().getResource("gold/documentation").getPath();
    }

    public void testGenerateDocHandlesNull() {
        final String text = provider.generateDoc(null, null);

        assertNull("Should return null rather than throwing exception", text);
    }

    public void testGenerateDocForPublicClassVariable() {
        this.myFixture.configureByText("example.d", "class User { int id; string name;}");

        // class User { int id; string name;}
        //                  ^
        myFixture.getEditor().getCaretModel().moveToOffset(17);

        final PsiElement docElement = DocumentationManager.getInstance(getProject())
            .findTargetElement(myFixture.getEditor(), myFixture.getFile());

        final String text = provider.generateDoc(docElement, null);

        assertNotNull(text);
        assertTrue(text.contains("symbol"));
        assertTrue(text.contains("public"));
        assertFalse(text.contains("private"));
    }

    public void testGenerateDocForPrivateClassVariable() {
        this.myFixture.configureByText("example.d", "class User { int id; private string name;}");

        // class User { int id; private string name;}
        //                                      ^
        myFixture.getEditor().getCaretModel().moveToOffset(37);

        final PsiElement docElement = DocumentationManager.getInstance(getProject())
            .findTargetElement(myFixture.getEditor(), myFixture.getFile());

        final String text = provider.generateDoc(docElement, null);

        assertNotNull(text);
        assertTrue(text.contains("symbol"));
        assertFalse(text.contains("public"));
        assertTrue(text.contains("private"));
    }

    public void testGenerateDocForGlobalInt() {
        this.myFixture.configureByText("int.d", "int x = 0;");

        // int x = 0;
        //     ^
        myFixture.getEditor().getCaretModel().moveToOffset(4);

        final PsiElement docElement = DocumentationManager.getInstance(getProject())
            .findTargetElement(myFixture.getEditor(), myFixture.getFile());

        final String text = provider.generateDoc(docElement, null);

        assertNotNull(text);
        assertTrue(text.contains("symbol"));
        assertTrue(text.contains("public"));
        assertFalse(text.contains("private"));
        assertTrue(text.contains("static"));
    }

    /*
     * Run a test that uses the "gold/documentation/example.d" file
     */
    public void testGenerateDocForFileBasedExample() {
        this.myFixture.configureByFile("example.d");

        // put the caret on the doSomething() function in the source file
        myFixture.getEditor().getCaretModel().moveToOffset(40);

        final PsiElement docElement = DocumentationManager.getInstance(getProject())
            .findTargetElement(myFixture.getEditor(), myFixture.getFile());

        final String text = provider.generateDoc(docElement, null);

        assertNotNull(text);
        assertTrue(text.contains("symbol"));
        assertTrue(text.contains("public"));
        assertFalse(text.contains("private"));
        assertTrue(text.contains("static"));
    }
}

File: src/test/kotlin/io/github/intellij/dlanguage/features/DDocumentationProviderTest.kt
package io.github.intellij.dlanguage.features

import com.intellij.codeInsight.documentation.DocumentationManager
import com.intellij.psi.util.startOffset
import com.intellij.testFramework.fixtures.LightPlatformCodeInsightFixture4TestCase
import io.github.intellij.dlanguage.psi.named.DLanguageClassDeclaration
import io.github.intellij.dlanguage.psi.named.DLanguageFunctionDeclaration
import org.junit.Test

class DDocumentationProviderTest : LightPlatformCodeInsightFixture4TestCase() {

    private var provider: DDocumentationProvider? = null

    @Throws(Exception::class)
    public override fun setUp() {
        super.setUp()
        provider = DDocumentationProvider()
    }

    override fun getTestDataPath(): String {
        return this.javaClass.classLoader.getResource("gold/documentation")!!.path
    }

    @Test
    fun testGetUrlForHandlesSingleImport() {
        myFixture.configureByText("example.d", "import std.typecons;")

        val elementAndContext = DocumentationManager.getInstance(project)
            .findTargetElementAndContext(myFixture.editor, 13, myFixture.file)!!

        val result = provider!!.getUrlFor(elementAndContext.first, elementAndContext.second)
        assertEquals("https://dlang.org/phobos/std_typecons.html", result[0])
    }

    @Test
    fun testGetUrlForHandlesNullAndNonImportStatement() {
        myFixture.configureByText("example.d", "class User { int id; string name;}")

        val elementAndContext = DocumentationManager.getInstance(project)
            .findTargetElementAndContext(myFixture.editor, 17, myFixture.file)!!

        assertEmpty("Should return empty list rather than throwing exception", provider!!.getUrlFor(null, null))
        assertEmpty("Should return empty list rather than throwing exception", provider!!.getUrlFor(elementAndContext.first, null))
        assertEmpty("Should return empty list rather than throwing exception", provider!!.getUrlFor(null, elementAndContext.second))
        assertEmpty("Should return empty list rather than throwing exception", provider!!.getUrlFor(elementAndContext.first, elementAndContext.second))
    }

    @Test
    fun testGenerateDocHandlesNull() {
        val text = provider!!.generateDoc(null, null)
        assertNull("Should return null rather than throwing exception", text)
    }

    @Test
    fun testGenerateDocForPublicClassVariable() {
        myFixture.configureByText("example.d", "class User { int id; string name;}")

        // class User { int id; string name;}
        //                  ^
        myFixture.editor.caretModel.moveToOffset(17)
        val docElement = DocumentationManager.getInstance(project)
            .findTargetElement(myFixture.editor, myFixture.file)
        val text = provider!!.generateDoc(docElement, null)
        assertNotNull(text)
        assertTrue(text!!.contains("int"))
        assertTrue(text.contains("id"))
    }

    @Test
    fun testGenerateDocForPrivateClassVariable() {
        myFixture.configureByText("example.d", "class User { int id; private string name;}")

        // class User { int id; private string name;}
        //                                      ^
        myFixture.editor.caretModel.moveToOffset(37)
        val docElement = DocumentationManager.getInstance(project)
            .findTargetElement(myFixture.editor, myFixture.file)
        val text = provider!!.generateDoc(docElement, null)
        assertNotNull(text)
        //assertTrue(text.contains("private")) XXX it’s private, so we can display it
        assertTrue(text!!.contains("string"))
        assertTrue(text.contains("name"))
    }

    @Test
    fun testGenerateDocForGlobalInt() {
        myFixture.configureByText("int.d", "int x = 0;")

        // int x = 0;
        //     ^
        myFixture.editor.caretModel.moveToOffset(4)
        val docElement = DocumentationManager.getInstance(project)
            .findTargetElement(myFixture.editor, myFixture.file)
        val text = provider!!.generateDoc(docElement, null)
        assertNotNull(text)
        assertTrue(text!!.contains("int"))
        assertTrue(text.contains("x"))
    }

    /*
     * Run a test that uses the "gold/documentation/example.d" file
     */
    @Test
    fun testGenerateDocForFileBasedExample() {
        myFixture.configureByFile("example.d")

        val doSomethingMethod = myFixture.findElementByText("doSomething", DLanguageFunctionDeclaration::class.java)

        // put the caret on the doSomething() function in the source file
        myFixture.editor.caretModel.moveToOffset(doSomethingMethod!!.identifier!!.startOffset)
        val docElement = DocumentationManager.getInstance(project)
            .findTargetElement(myFixture.editor, myFixture.file)
        val text = provider!!.generateDoc(docElement, null)
        assertNotNull(text)
        //assertTrue(text!!.contains("static"))
        assertTrue(text!!.contains("void"))
        assertTrue(text.contains("doSomething"))
        assertTrue(text.contains("()"))
        assertTrue(text.contains("This is the method documentation."))
    }

    @Test
    fun testGenerateDocForClassBasedExample() {
        myFixture.configureByFile("example.d")

        val myCodeClass = myFixture.findElementByText("MyCode", DLanguageClassDeclaration::class.java)

        // put the caret on the doSomething() function in the source file
        myFixture.editor.caretModel.moveToOffset(myCodeClass!!.identifier!!.startOffset)
        val docElement = DocumentationManager.getInstance(project)
            .findTargetElement(myFixture.editor, myFixture.file)
        val text = provider!!.generateDoc(docElement, null)
        assertNotNull(text)

        assertTrue(text!!.contains("This is the CLASS documentation"))
    }


    @Test
    fun testGenerateDocWithEmptyCodeSnippet() {
        myFixture.configureByText("example.d", """
            /**
             Contains empty code snippet
             ~~~ ~~~*/
             class I {}
        """.trimIndent())

        val myCodeClass = myFixture.findElementByText("I", DLanguageClassDeclaration::class.java)

        // put the caret on the doSomething() function in the source file
        myFixture.editor.caretModel.moveToOffset(myCodeClass!!.identifier!!.startOffset)
        val docElement = DocumentationManager.getInstance(project)
            .findTargetElement(myFixture.editor, myFixture.file)
        val text = provider!!.generateDoc(docElement, null)
        assertNotNull(text)

        assertTrue(text!!.contains("Contains empty code snippet"))
    }
}
